### ISSUE TITLE
problem truncate

### ISSUE THREAD

**Issue by 8carlosf** [baskerville/xtitle#5]
for demonstration do

``` ini
xtitle -s -t 8
```

or

``` ini
xtitle -s -t -8
```

and then open a terminal with this name

``` ini
urxvt -name óóóóóó
```

the xtitle output will be "óóÃ..."

so,
1 - the chars are not counted well (thats normal, because char only has 256 values, non-ASCII characters take more than that), this is no big deal (in my opinion)
2 - truncating a char in half may cause problems (and it will cause problems in i3bar, it will mess up your json uft-8, it found the hard way)

I hope someone was suggestions to solve this problem (maybe add some filters to the  @timaa2k escape mode?!), can't think of one right now. I don't even understand well why this is a problem to the json or utf-8.

**Comment by 8carlosf** [comment]
after some debug I found that every (non-ascii char uses 2 normal chars and the first one is always Ã or -61 (numeric) also know as multi-character constant.
so, after some time, my suggestion is the following
``` ini
if (truncate) {
		unsigned int n = abs(truncate);
		if (strlen(title) > (size_t)n) {
			if (truncate > 0) {
				if (n > 3) {
					title[n-1] = '.';
					title[n-2] = '.';
					if (title[n-4] != -61)
						title[n-3] = '.';
				}
				title[truncate] = '\0';
			} else {
				title = title + strlen(title) + truncate;
				if (n > 3) {
					title[0] = '.';
					title[1] = '.';
					if (title[2] != -61)
						title[2] = '.';
				}
			}
		}
	}
}
```

this will put only to ".." instead of "..." in some cases, it was the simple way of doing it.

so, explaining:
in the right truncate, if the last char a multi-char constant (-61) it would break the uft-8, so instead I will leave the next char in (don't break the pair).
in the left truncate (negative values), I check if I'm putting a . in a multi-char constant, again, this way I will not break the pair.
I don't know if some utf-8 have more than 2 values in a row.. I hope it doesn't.. anyway, tell me what you think about this (I can submit a pull request if you want)

**Comment by 8carlosf** [comment]
my solution doesn't work for non latin letters.. so, still broken.. any ideas on this?
this is a bad problem for people using i3 (i3bar). I will have a look at uft-8 in more detail when I can...

**Comment by baskerville** [comment]
I'll probably have to use a `wchar_t` (and call `setlocale`) to fix this properly.

It will only work when the title comes from a `_NET_WM_NAME` property though, because when it comes from `WM_NAME`, the encoding is *Compound Text*...

**Comment by 8carlosf** [comment]
well, my solution (this time I'm 99% sure it works!)
``` ini
xtitle -s -t -111 | xargs -L 1 -I {} sh -c "echo '{}' | iconv -c -f utf-8 -t utf-8"
```
I would suggest a recommendation in the -t option for this solution, or at least an alert for the problem.
And also, if you could change the MAXLEN to 512 or 1024 (both are low amounts of RAM) it would be nice, because if you have some title in greek (or other language with less usual chars) the real utf-8 length of the string with 256 bytes is far less than that..

sorry for the previous code changes, this is a harder problem to solve than I anticipated.

**Comment by d125q** [comment]
@baskerville Is there no way to convert from Compound Text to UTF-8?

**Comment by baskerville** [comment]
I don't know.

**Comment by cmellwig** [comment]
Using `wchar_t` is the proper way. Sadly I don't have time to implement. I did this quick and dirty fix, because the error was crashing my i3bar: https://github.com/cmellwig/xtitle/commit/100eb2258358044cd668cc7370935a5d4c48aadc

**Comment by d125q** [comment]
> Using `wchar_t` is the proper way.

I would say that using [ICU](http://site.icu-project.org/) is the proper way.  It even provides a way to convert from Compound Text to other encodings.

**Comment by baskerville** [comment]
This is fixed by 942bfdd.

And as of 69e4780, `xterm -T ☺` will work as expected.
